### DISCUSSION TITLE
this 키워드란 무엇인가요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#95] [this]
- this 키워드란 무엇인가요?
    - this 바인딩이란 무엇인가요?
    - this가 필요한 이유가 뭔가요?

**Comment by publdaze** [comment]
> - this 키워드란 무엇인가요?
- this란 자신이 속한 객체나 자신이 생성할 인스턴스를 가리키는 자기 참조 변수
> - this 바인딩이란 무엇인가요?
- this 식별자에 this가 가리킬 객체를 연결하는 것
- 함수 호출 방식에 따라 this에 바인딩될 값이 동적으로 결정
> - this가 필요한 이유가 뭔가요?
- 객체의 메서드가 자신이 속한 객체의 프로퍼티를 참조하기 위해 필요

**Comment by pipisebastian** [comment]
> * this 키워드란 무엇인가요?
- 자신이 속한 객체, 자신이 생성할 인스턴스를 가리키는 자기 참조 변수입니다.
>   * this 바인딩이란 무엇인가요?
- 실행컨텍스트의 this에 참조할 객체를 연결하는 것입니다. 함수가 호출되는 시점에 this는 동적으로 정해집니다.
>   * this가 필요한 이유가 뭔가요?
- **자신이 속한 객체의 프로퍼티를 참조하기 위해 필요** (publdaze)
- 만약 **객체**를 예로 들면, 객체에서는 객체 식별자를 재귀적으로 참조할 수 있다. 
- 그러나 이 방식은 먼저, 객체 식별자 이름을 바꾼다면, 이를 사용하는 모든 코드를 수정해야하기에, 유지보수가 어려워집니다. 
- 또한, 만약 A 객체의 메서드를, B 객체에서도 참조하고 싶을때, 해당 메서드가 A 객체의 프로퍼티를 참조하고 있다면, 이는 B 객체의 메서드로 넣더라도, A 객체를 참조하고 있으므로, 원하는 동작을 기대할 수 없습니다.
```js
const circle = {
    radius: 5,
    getDiameter() {
        return circle.radius * 2; 
    }
};

const newCircle = {
    radius: 10,
    getDiameter: circle.getDiameter // circle의 메서드 참조
};

console.log(newCircle.getDiameter()); // 20을 기대하지만, 10이 출력됨
```
- 또한 생성자함수로 생성하는 인스턴스를 예로 들면, 생성자 함수를 정의하는 시점에서는, 자신이 누군지 알 수가 없습니다. (new로 인스턴스를 만든 뒤, 그 인스턴스를 가리켜야 하기 때문)
- 이러한 이유로, 자신을 가리키는 식별자, this가 필요합니다
